### PR TITLE
Redirect stdout and stderr to client

### PR DESCRIPTION
**Pull request by tinmarino**
Instead of getting them in the old `sys.stdout`
Fix part1 of issue #2 
ipy_thread

### CODE REVIEW

File: kernel.py [hunk 1]
```diff
@@ -228,7 +251,7 @@ def _thread_loop(self):
         loop.add_callback(self._start_kernel)
         try:
             loop.start()
-        except KeyboardInterrupt:
+        except (KeyboardInterrupt, SystemExit):
```
Comment (albertz) [inline kernel.py]: Why? In what cases will a `SystemExit` raised here? And if it is raised, then shouldn't it actually handle it?

File: kernel.py [hunk 2]
```diff
@@ -98,6 +101,23 @@ def __init__(self, connection_filename="kernel.json", connection_fn_with_pid=Tru
             logger.addHandler(logging.StreamHandler(sys.stdout))
         self._logger = logger
 
+    def _init_io(self):
+        """
+        Redirect stdout to iopub socket
+        call me after loging conection file
```
Comment (albertz) [inline kernel.py]: There are some typos here.

File: kernel.py [hunk 3]
```diff
@@ -8,6 +8,8 @@
 from tornado import ioloop
 from ipykernel.ipkernel import IPythonKernel, ZMQInteractiveShell
 
+from ipykernel.iostream import OutStream
```
Comment (albertz) [inline kernel.py:10]: Why leave the empty line before this?

File: kernel.py (hunk 1 above)
Comment (albertz) [reply kernel.py]: Is this related to `keepkernel_on_exit`?

File: kernel.py [hunk 4]
```diff
@@ -211,6 +232,8 @@ def _start_kernel(self):
         self._create_kernel()
 
         self._logger.info("IPython: Start kernel now. pid: %i, thread: %r" % (os.getpid(), threading.current_thread()))
+        self._init_io()
```
Comment (albertz) [inline kernel.py]: I don't think it's a good idea to do that by default. The intention of this project (background-zmq-ipython) is that it runs in the background. It should not change and affect the main program (like changing `sys.stdout`).

But I see your use case for that. So better just make that an option for `IPythonBackgroundKernelWrapper`.

**Review by albertz — CHANGES_REQUESTED** [review]
See comments.

File: kernel.py (hunk 1 above)
Comment (tinmarino) [reply kernel.py]: Short story: It is obsolete, from a previous implementation I did not commit

I did not manage not trigger it. I could not trigger the KeyboardInterrupt either.
Yes it is for exiting the kernel, with a `sys.exit` <- `exit(0)`.
I did not see the `atexit.register` so I wanted to clean the stdio redirection there.

**Comment by tinmarino** [comment]
I think, by default, every user want to see his own output.
The last implementation is changing the output only of the current thread: 54970642
Note: if someone else is tampering with std streams (ex: an other async kernel), it should work flawless : the new proxy will call his one if not in the current kernel thread.

What do you think of it? If you don't like it, may you tell me the name of the witch parameter ? Something like `redirect_stream=False`.

File: kernel.py (hunk 1 above)
Comment (tinmarino) [reply kernel.py]: Fix: 572d8f5

File: kernel.py (hunk 2 above)
Comment (tinmarino) [reply kernel.py]: Fix: a940381f0a

File: kernel.py (hunk 3 above)
Comment (tinmarino) [reply kernel.py:10]: Fix: a940381f0a

File: kernel.py (hunk 4 above)
Comment (tinmarino) [reply kernel.py]: Fix: 54970642 but not a parameter yet (waiting for your opinion)

Comment (albertz) [reply kernel.py]: This is somewhat hacky, and could have unintended side effects. I really would make this only optional. Some option like `redirect_stdio` which by default is `False`, and if `True`, it uses your code.

Comment (albertz) [reply kernel.py]: Also, I'm actually not sure what the nicer behavior is. Maybe for `redirect_stdio=True`, better really redirect all stdio, not just from this thread. Or maybe only if `redirect_stdio="this_thread_only"` or so, not sure. I would leave that away for now, and either redirect it not all all, or fully.

Comment (tinmarino) [reply kernel.py]: Made it optional (default disabled) : 37c526d
If enabled, it redirects the current thread <-  I am personally interested in only the thread output (the `print` from the controlling console).
I tested with different client and it behaves exactly as I need with `redirect_stdio=True`.

By  the way I agree that the default ipykernel implementation "should" look like yours. But that won't be the case for years. So ... thanks again for your code.

Comment (albertz) [reply kernel.py]: Ok.

**Review by albertz — APPROVED** [review]

**Comment by albertz** [comment]
Thanks. Squashed and merged.

**Comment by tinmarino** [comment]
Thank You Albertz, you wrote a pretty good kernel. And were really fast to review my PRs.

**Pull request merged**